google/cloud/spanner/row_parser.h

```diff
@@ -0,0 +1,165 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_SPANNER_ROW_PARSER_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_SPANNER_ROW_PARSER_H_
+
+#include "google/cloud/spanner/row.h"
+#include "google/cloud/spanner/value.h"
+#include "google/cloud/spanner/version.h"
+#include "google/cloud/status.h"
+#include "google/cloud/status_or.h"
+#include <array>
+#include <tuple>
+#include <utility>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+
+/**
+ * A RowParser takes a range of `Value` objects and a list of types, and it
+ * parses the values into a range of `Row` objects with the specified types.
+ *
+ * Callers should create instances of this class with the `MakeRowParser()`
+ * function (defined below), which takes an input range and a list of types
+ * that correspond to each row. The returned `RowParser` can be iterated,
+ * producing a range of `StatusOr<Row<Ts...>>`. The `StatusOr` is needed in
+ * case the input `Value` objects cannot be parsed to the requested type.
+ *
+ * Example:
+ *
+ *     std::vector<Value> const values = {
+ *         Value(true), Value(0),  // Row 0
+ *         Value(true), Value(1),  // Row 1
+ *         Value(true), Value(2),  // Row 2
+ *         Value(true), Value(3),  // Row 3
+ *     };
+ *     for (auto row : MakeRowParser<bool, std::int64_t>(values)) {
+ *       if (!row) {
+ *         // handle error
+ *       }
+ *       bool b = row->get<0>();
+ *       std::int64_t i = row->get<1>();
+ *     }
+ *
+ * @tparam ValueIterator the iterate for iterating `Value` objects.
+ * @tparam Ts... the column types for each `Row`
+ */
+template <typename ValueIterator, typename... Ts>
+class RowParser {
+  class RowIterator;  // Defined in private section below.
+
+ public:
+  using iterator = RowIterator;
+
+  /// Constructs a `RowParser` for the given range of `Value`s.
+  template <typename Range>
+  explicit RowParser(Range&& range)
+      : it_(std::begin(std::forward<Range>(range))),
+        end_(std::end(std::forward<Range>(range))) {}
+
+  /// Copy and assignable.
+  RowParser(RowParser const&) = default;
+  RowParser& operator=(RowParser const&) = default;
+  RowParser(RowParser&&) = default;
+  RowParser& operator=(RowParser&&) = default;
+
+  /// Returns the begin iterator.
+  iterator begin() { return iterator(it_, end_); }
+
+  /// Returns the end iterator.
+  iterator end() { return iterator(end_, end_); }
+
+ private:
+  // The RowParser's interator type. Models InputIterator so it can only
+  // iterate forward and can only do so once.
+  class RowIterator {
```

Comment:
Why not just a public class called `iterator`?  In any case, this is part of the public interface, it needs documentation like any other class...  Either here or in line 66 something like this maybe?

```C++
/// A single-pass input iterator that coalesces multiple `Value` results into a `RowType`.
```

Reply:
Good point. Done. This is better because the iterator's member aliases and stuff really are public API points.

FTR, the reason it was private was to prevent others from directly creating instances of this iterator. But I can do the same thing by making the class public but making its c'tor private (and friending `RowParser`), which I've done.